torch/csrc/jit/export.cpp

```diff
@@ -571,8 +645,22 @@ void GraphEncoder::EncodeTensor(
     tensor_proto->set_raw_data("__EXTERNAL");
   } else {
     AT_ASSERT(t.is_contiguous());
-    tensor_proto->set_raw_data(std::string(
+    size_t tensorSize = static_cast<size_t>(std::accumulate(std::begin(tensor.sizes()), 
+                        std::end(tensor.sizes()), static_cast<int64_t>(1), std::multiplies<int64_t>()));
+    if (use_large_model_format && tensorSize > ParamSizeThresholdForExternalStorage) {
```

Comment:
so if use_large_model_format is enabled but tensors sizes are smaller than the threshold, we will save them in the model.

In the scenario where the model has a very big number of parameters, and each of them have a size that is slightly below ParamSizeThresholdForExternalStorage, we will be exporting the model as a normal non-large model, which could potentially be an issue.

Do we want to warn the user of why they are not saved in the external storage?
Do we want to give the user the flexibility to choose this threshold value (or to choose between enabling/disabling a fixed threshold)?

This is a very special case, so we probably don't need to complicate things for now, but I am just wondering how we chose the value of ParamSizeThresholdForExternalStorage.

Reply:
There are a lot of different choices here. If we create one file for each parameter, however small, then we have a large number of files. On the other hand, creating a single file with all parameters will, in most cases, create a huge file, which is also not ideal. This is a middle path. I chose the threshold after looking at different models and choosing a sweet spot for the threshold. 
Yes, it is theoretically possible that there's a model which has all parameters below the threshold (currently set at 1024 elements) but still exceeds the 2GB limit, but that could happen for models with really extremely large topologies. Frankly, I don't expect any practical models to be like that. Most models have weight tensor that easily exceed 1024 elements, and therefore will write an external file is large model format is enabled.

Among other options, I have considered giving more control to the user, but at this point I would want to keep things simple and not create more API parameters. I do not expect the user to have an easy way to judge what's the best threshold in most cases. Also, I don't think a warning here will help much, and may create unwanted noise. In the doc for this argument, I do mention that only "some" parameters are stored in external files (didn't want to be more specific about the logic as it may change in the future). But your point is valid, and this is a first cut. In the future, if a better design comes out of the user experience, I expect that we will update this logic.